spec/model_extension_spec.rb

```diff
@@ -0,0 +1,60 @@
+# spec/model_extension_spec.rb
+
+RSpec.describe Readyset::ModelExtension do
+  describe '.readyset_query' do
+    subject { Cat.readyset_query(name, body) }
+
+    context 'when the body is not callable' do
+      let(:body) { 'I am not callable' }
+      let(:name) { :find_by_name_cached }
+
+      it 'raises an ArgumentError' do
+        expect { subject }.to raise_error(ArgumentError)
+      end
+    end
+
+    context "when the query's name would overwrite a predefined class method" do
+      let(:body) { ->(name) { Cat.where(name: name) } }
+      let(:name) { :all }
+
+      it 'raises an ArgumentError' do
+        expect { subject }.to raise_error(ArgumentError)
+      end
+    end
+
+    context "when the query's name would overwrite a predefined ActiveRecord::Relation method" do
+      let(:body) { ->(name) { Cat.where(name: name) } }
+      let(:name) { :load }
+
+      it 'raises an ArgumentError' do
+        expect { subject }.to raise_error(ArgumentError)
+      end
+    end
+
+    context 'when the body is callable and the name is valid' do
+      let(:name) { :find_by_name_cached }
+      let(:body) { ->(name) { Cat.where(name: name) } }
+
+      before do
+        # Add a few records to the database that is acting as our fake ReadySet instance
+        Readyset.route(prevent_writes: false) do
+          Cat.create!(name: 'whiskers')
+          Cat.create!(name: 'fluffy')
+          Cat.create!(name: 'tails')
+        end
+
+        subject
+      end
+
+      it "defines a method on the model's class" do
+        expect(Cat).to respond_to(name)
+      end
+
+      it 'defines a method that returns query results from ReadySet' do
+        results = Cat.find_by_name_cached('whiskers')
+        expect(results.size).to eq(1)
+        expect(results.first.name).to eq('whiskers')
+      end
```

Comment:
I'm having some difficulty with understanding what does this do? 

I feel like this spec describes how it works, but not necessarily what it's doing from a user's perspective.

> defines a method (:find_by_name_cached)
The user writes a method (`name = :find_by_name_cached`) to be called on `Cat`
> returns query results from ReadySet
Said method performs the query and returns the records from ReadySet?

Reply:
Yup, `.readyset_query` defines a new method on `self` (an ActiveRecord model) using the query in the given lambda such that the query is routed to ReadySet. It's similar  ActiveRecord's `scope` in that it defines a new method that wraps the query given as a parameter